### PR TITLE
ROX-14660: Mutiplex NodeInventory in the same Node worker queue

### PR DESCRIPTION
**Pull request by jvdm**
## Description

Force `NodeInventory` sensor events to the same worker queue as `Node` events. Since [Node](https://github.com/stackrox/stackrox/blob/master/sensor/kubernetes/listener/resources/node.go#L59) and [NodeInventory](https://github.com/stackrox/stackrox/blob/master/sensor/common/compliance/node_inventory_handler_impl.go#L122) use the Node ID as the event message ID, they will [land on the same shard](https://github.com/stackrox/stackrox/blob/master/central/sensor/service/connection/worker_queue.go#L49-L48) and effectively be processed in the same go routine and never run concurrently.

We want those two event types to serialize because they are complementary. While both event pipelines change the Node database, the Node event must only change the node metadata fields. In contrast, the NodeInventory event only needs to update the [NodeScan scan](https://github.com/stackrox/stackrox/blob/5da844ac7bb20f311ef5c20256b301b6c5ceb459/proto/storage/node.proto#L50). But, the database schema and model APIs don't offer a way to do that.

So the pipelines rely on READ-TO-WRITE to read the current state, modify their target fields, leave the other fields untouched, and write the whole Node back to the database. This approach is generally discouraged as it can lead to inconsistencies and conflicts if the two events run and try to read and write the same data simultaneously. **Nonetheless, instead of locking or transaction management to ensure consistency, we use a more straightforward approach and serialize the events.** There is no performance gain in having these two pipelines run concurrently.

## Checklist
- [x] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [x] ~Evaluated and added CHANGELOG entry if required~ N/A
- [X] ~Determined and documented upgrade steps~ N/A
- [X] ~Documented user-facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~ N/A

## Testing Performed

This change has been deployed to an Openshift-4 cluster, and node and node inventory were observed to work as expected.

Two tests were performed.

### Metrics

Moreover, I took the opportunity of the fact that the deduping queues emit metrics per queue type. Before the change metrics found were:

```
rox_central_sensor_event_queue{Operation="Add",Type="Node"} 6
rox_central_sensor_event_queue{Operation="Add",Type="NodeInventory"} 6
```

After the change:

```
rox_central_sensor_event_queue{Operation="Add",Type="Node"} 12
```

This is expected: NodeInventory was mapped to Node, yielding a total count of 12 (6 Node + 6 NodeInventory).

### Manual sleeps

1. I added a conditional sleep in the Node pipeline for one of the nodes in my cluster:
   ```
   sensor/service/pipeline/nodes: 2023/03/07 22:48:10.441183 pipeline.go:105: Info: JVDM: Sleep for only 5 min, go trigger node inventory now: Name=jvdm-argilo-47bkl-master-0.c.srox-temp-dev-test.internal ID=1ce86640-8407-4e01-aed6-665da2362936
   ```

2. Then I triggered node inventory by killing compliance:
   ```
   for p in $(kubectl -n stackrox get po -l app=collector --output=jsonpath='{.items[*].metadata.name}')
       do s=$(( 1 + $RANDOM % 5 ))
       echo $s
       sleep $s
       kubectl -n stackrox exec $p -c compliance -- kill 1
   done
   ```

3. Finally, observed the Node Inventory being processed only after the 5 minute interval:
   ```
   sensor/service/pipeline/nodeinventory: 2023/03/07 22:53:14.492277 pipeline.go:65: Info: Received NodeInventory for Node name='jvdm-argilo-47bkl-master-0.c.srox-temp-dev-test.internal' ID='1ce86640-8407-4e01-aed6-665da2362936'
   ```

### CODE REVIEW

**Comment by jvdm** [comment]
@vikin91 This effectively disables the race between the two pipelines. It simplifies since we don't have to deal with the race at the DB level with conditional checks. We still need to ensure Nodes won't overwrite node scanning scans. Expect another PR for that. :pray:

File: central/sensor/service/connection/sensorevents.go [hunk 1]
```diff
@@ -12,15 +13,21 @@ import (
 	"github.com/stackrox/rox/generated/internalapi/central"
 	"github.com/stackrox/rox/pkg/concurrency"
 	"github.com/stackrox/rox/pkg/reflectutils"
+	"github.com/stackrox/rox/pkg/sync"
 	"github.com/stackrox/rox/pkg/utils"
 )
 
 const workerQueueSize = 16
```
Comment (vikin91) [inline central/sensor/service/connection/sensorevents.go:20]: Does this need to be doubled for the queue holding nodes and nodeInventories now?

**Review by vikin91 — COMMENTED** [review]
Great idea! Thanks for experimenting with it and proposing it!

I am generally on the ✅ side, but I would like to first understand it better, so I left few questions.

**Comment by vikin91** [comment]
> @vikin91 This effectively disables the race between the two pipelines. It simplifies since we don't have to deal with the race at the DB level with conditional checks. We still need to ensure Nodes won't overwrite node scanning scans. Expect another PR for that. 🙏

Great! Would that be still done with not calling the Scanner from the nodes/pipeline when RHCOS is detected?

File: central/sensor/service/connection/sensorevents.go [hunk 2]
```diff
@@ -12,15 +13,21 @@ import (
...
+var (
+	deploymentEventType = reflectutils.Type((*central.SensorEvent_Deployment)(nil))
+	nodeEventType       = reflectutils.Type((*central.SensorEvent_Node)(nil))
+)
 
 type sensorEventHandler struct {
-	typeToQueue map[string]*workerQueue
+	// workerQueues are keyed by central.SensorEvent type names.
+	workerQueues      map[string]*workerQueue
+	workerQueuesMutex sync.Mutex
```
[the remark this answers is not in the document]
Comment (connorgorman) [reply central/sensor/service/connection/sensorevents.go:30]: Mutexes do not need to be initialized

File: central/sensor/service/connection/sensorevents.go [hunk 3]
```diff
@@ -86,12 +100,17 @@ func (s *sensorEventHandler) addMultiplexed(ctx context.Context, msg *central.Ms
...
-		s.typeToQueue[typ] = queue
-		go queue.run(ctx, s.stopSig, s.handleMessages)
+		concurrency.WithLock(&s.workerQueuesMutex, func() {
+			queue = s.workerQueues[eventType]
+			if queue == nil {
+				queue = newWorkerQueue(workerQueueSize, stripTypePrefix(eventType), s.injector)
+				s.workerQueues[eventType] = queue
+				go queue.run(ctx, s.stopSig, s.handleMessages)
+			}
+		})
```
[the remark this answers is not in the document]
Comment (connorgorman) [reply central/sensor/service/connection/sensorevents.go:113]: I actually think this locking pattern is problematic even though its very unlikely to happen in reality. Could have someone reading s.workerQueues and writing to s.workerQueues which will trigger a concurrent map read/write. We should probably wrap the s.workerQueues part in a RLock, and then can get an exclusive lock for the few times we actually modify s.workerQueues. Should have practically no perf implcations

[the remark this answers is not in the document]
Comment (connorgorman) [reply central/sensor/service/connection/sensorevents.go:113]: @jvdm I know this pattern is used elsewhere and we should probably fix it there as well

File: central/sensor/service/connection/sensorevents.go [hunk 4]
```diff
@@ -66,9 +73,16 @@ func (s *sensorEventHandler) addMultiplexed(ctx context.Context, msg *central.Ms
...
 		case *central.SensorEvent_ReprocessDeployment:
-			typ = deploymentQueueKey
+			eventType = deploymentEventType
+		case *central.SensorEvent_NodeInventory:
+			// This will put both NodeInventory and Node events in the same worker queue,
+			// preventing events for the same Node ID to run concurrently.
+			eventType = nodeEventType
+			// Node and NodeInventory dedupe on Node ID. We use a different dedupe key for
+			// NodeInventory because the two should not dedupe between themselves.
+			msg.DedupeKey = fmt.Sprintf("%s:%s", "NodeInventory", msg.GetDedupeKey())
```
[the remark this answers is not in the document]
Comment (jvdm) [reply central/sensor/service/connection/sensorevents.go]: Fix at de30e77c6c

File: central/sensor/service/connection/sensorevents.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (jvdm) [reply central/sensor/service/connection/sensorevents.go]: Fix at e84f902161

I decided to update one place I know this pattern is also used.

File: central/sensor/service/connection/sensorevents.go (hunk 1 above)
Comment (jvdm) [reply central/sensor/service/connection/sensorevents.go:20]: Could you please clarify what specifically you want to double for the worker queues? Also, I assume you are talking about the deduping queues for Node and NodeIventory if you could clarify? Need that information to be able to answer your question.

File: central/sensor/service/connection/sensorevents.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (jvdm) [reply central/sensor/service/connection/sensorevents.go]: To complement Connor's: Mutexes with zero values are ready to go. And actually, by assigning `sync.Mutex{}` you are effectively doing that. Golang always initializes variables to zero, so `var m sync.Mutex` is already "initialized".

**Comment by jvdm** [comment]
> Great! Would that be still done with not calling the Scanner from the nodes/pipeline when RHCOS is detected?

Yes, that is what I meant by "We still need to ensure Nodes won't overwrite node scanning scans. Expect another PR for that.".

File: central/sensor/service/connection/sensorevents.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (vikin91) [reply central/sensor/service/connection/sensorevents.go]: Sorry, you are right. I must have confused it with something else.

> The zero value for a Mutex is an unlocked mutex

File: central/sensor/service/connection/sensorevents.go (hunk 1 above)
Comment (vikin91) [reply central/sensor/service/connection/sensorevents.go:20]: Sure, here is my thinking. We used to have 2 separate queues for node and nodeInventories - each of capacity 16. Now, we place both kind of messages into a single queue of size 16, which means that the capacity of 16 needs to be shared between the nodes and nodeInventories, which (on average) reduces the capacity to 8 messages of one type.
Is this correct?

I do not know what are the consequences of having less capacity in the queue, so I would also like to understand what would be the consequence of setting it to, for example, 1.

Comment (jvdm) [reply central/sensor/service/connection/sensorevents.go:20]: Oh, I see now. The `workerQueueSize` is the number of shards inside the main worker queue. Each worker queue has 16 shards; the shard key is the resource ID. That means each resource is handled serialized in the same Q, but multiple resources can be handled concurrently. The name `workerQueueSize` is just a lousy name IMO: it should be `shardCount` or similar.

**Pull request merged**